File: python/98_Validate_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @return {boolean}
    def isValidBST(self, root):
        if not root:
            return True
        queue=[]
        while root:
            #if root.left and root.left.val>root.val:
            #    return False
            queue.append(root)
            root=root.left
        pre=node=queue.pop()
        if node.right:
            node=node.right
            while node:
                queue.append(node)
                node=node.left
        while queue:
            node=queue.pop()
            if pre.val>=node.val:
                return False
            pre=node
            if node.right:
                node=node.right
                while node:
                    queue.append(node)
                    node=node.left
        return True

File: Java/98_Validate_Binary_Search_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    // 树的中序遍历
    public boolean isValidBST(TreeNode root) {
        if(root==null)return true;
        LinkedList<TreeNode> queue = new LinkedList<>();
        TreeNode cur, pre=null;
        queue.add(root);
        while(root.left!=null){
            root=root.left;
            queue.add(root);
        }
        while(queue.size()>0){
            cur = queue.pollLast();
            if(pre!=null&&cur.val<=pre.val)return false;
            else{
                pre = cur;
                if(cur.right!=null){
                    cur=cur.right;
                    queue.add(cur);
                    while(cur.left!=null){
                        cur=cur.left;
                        queue.add(cur);
                    }
                }
            }
        }
        return true;
    }
}